hashintel/hash
# Handle external types in HASH pages

## 🌟 What is the purpose of this PR?



This PR assigns blocks on pages their actual block entity type ID when creating block links and block instances. This leverages the work done with the type fetcher to automatically fetch the type of the block without needing to explicitly create the block before creating the block entities.

I had to change some adjust some of the existing code around the `blockMap` in the frontend code and the type fetcher transport for it to work end-to-end. 

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1202805690238892/1203771979583945/f) _(internal)_

## 🚫 Blocked by





## 🔍 What does this change?




- Adds the component ID to BlockMeta map as a dependency to the prosemirror save step
- Change type fetcher's transport from Messagepack to JSON as we were seeing strange behavior with Messagepack (see https://hashintel.slack.com/archives/C03F7V6DU9M/p1683203634322339)
- Correctly implement the componentId to block entity type id mapping function
- Remove the createEntityType page action
- Bumps Rust BP Type System git dependency
- Bumps rooks to `7.14.1`

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:













### 📜 Does this require a change to the docs?




The changes in this PR:





### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:






## ⚠️ Known issues



## 🐾 Next steps



## 🛡 What tests cover this?



-

## ❓ How to test this?



1.  Checkout the branch, start the HASH app (external services and `yarn dev`)
1.  Sign in with any test user and create a page. Add a BP block to the page such as the kanban board or the code block
1.  Confirm that the blocks are working
1.  Confirm that the DB now contains new entity types that reflect the entity type IDs of the blocks
I used the following PG query:
```sql
select base_url, version, additional_metadata, properties from entity_editions ee
inner join entity_is_of_type et on et.entity_edition_id = ee.entity_edition_id
inner join ontology_id_with_metadata om on et.entity_type_ontology_id = om.ontology_id
where not starts_with(base_url, 'http://localhost')
```

## 📹 Demo


![image](https://user-images.githubusercontent.com/6988668/236265158-dde70e6c-722e-40a8-a368-4bba0db7af9d.png)
